### PR TITLE
Hooks refactor for Quiz component

### PR DESCRIPTION
**Pull request by rszeredi**
If you want to test without cloning the repo, a previous version is deployed at: https://rszeredi.github.io/countries-quiz/

Main changes in this PR:
- Refactoring the main (class-based) component <Quiz/> to a function component using hooks

BTW: this is my first real app (outside of course exercises), so there's probably a tonne that can be improved - all feedback is welcome!

--------------------------------------------------------------------------------------------------------
Here's a list of things I likely haven't implemented in the best way / you might have opinions on :

1. a)
	 Where to put functions now that they're no longer methods? (At the moment, I have half inside and half outside because wasn't sure)
   Outside the function-based component? Inside the component? 

	 1 b)
	 Is it bad practice for nested functions to rely on state variables (from the outer scope), or should they be turned into parameters?
	 Eg. handleAnswerSubmit uses currentQuestionIdx (which is a state variable). Is this ok, or is it instead better to have this as a separate function and pass in any necessary state variables and/or setState functions?

2. Updating state based on previous state. Is it guaranteed that the previous state will have already updated?
	 Eg. in updateScore function, replacing
	        this.setState((curSt) => ({ correct: curSt.correct + 1 })); 
		with 
		      setCorrect(currentCorrect + 1)
		     

3. Am I using useEffect in the correct way? (ie. as a replacement for componentDidMount)

4. Do I have too many state variables?

### CODE REVIEW

**Comment by alexrvincent** [comment]
> If you want to test without cloning the repo, a previous version is deployed at: https://rszeredi.github.io/countries-quiz/
> 
> Main changes in this PR:
> 
> * Refactoring the main (class-based) component  to a function component using hooks
> 
> BTW: this is my first real app (outside of course exercises), so there's probably a tonne that can be improved - all feedback is welcome!
> 
> Here's a list of things I likely haven't implemented in the best way / you might have opinions on :
> 
> 1. a)
>    Where to put functions now that they're no longer methods? (At the moment, I have half inside and half outside because wasn't sure)
>    Outside the function-based component? Inside the component?
>    1 b)
>    Is it bad practice for nested functions to rely on state variables (from the outer scope), or should they be turned into parameters?
>    Eg. handleAnswerSubmit uses currentQuestionIdx (which is a state variable). Is this ok, or is it instead better to have this as a separate function and pass in any necessary state variables and/or setState functions?
> 2. Updating state based on previous state. Is it guaranteed that the previous state will have already updated?
>    Eg. in updateScore function, replacing
>    this.setState((curSt) => ({ correct: curSt.correct + 1 }));
>    with
>    setCorrect(currentCorrect + 1)
> 3. Am I using useEffect in the correct way? (ie. as a replacement for componentDidMount)
> 4. Do I have too many state variables?

This looks great! Let me start by answering your direct questions, then I'll poke around some files and leave comments for things that stand out.

1. a) Depends what the purpose of the function is and where you're defining it😄 . In class components, you should keep functions closest to the places they're used the most (ie inside the component). General utility functions should definitely be broken out for sure. If the function is a callback (which it usually mostly is) and you're using hooks you can use something like `useCallback()` which will memoize the function instance to prevent it from being declared everytime the component renders 😄 .

   b). It's not bad per-se, as you'll need to run a setState() *somewhere* and breaking it out to a function makes sense. Personally I would for the following reasons,:
   - It'll be much easier to unit test and debug individually as you'll need pre-defined "state" to run the function.
   - You'll be able to reuse it if you create different types of form questions that follow different schema patterns   
   - If you use TypeScript you can add type safety 😄 
   
 2. Nothing in React *guarantees* a state value will be exactly what you want (as you can run into race conditions if you're setStating from elsewhere) but setStates are queued so they're safe to operate on in your setState callback assuming you're not doing anything that creates a side effect. The way you've done it is pretty common for class functions / setStates and works fine. I'd refactor it using state hooks so you're only ever setting what you want though.
 
 3. The way you're using `useEffect()` as a one-time componentDidMount refactor is correct, by adding an empty dependency array it'll only run once right after it mounts. There is a small difference between it and componentDidMount but none you should worry about unless you run into an issue. I'll comment on the line item too.
 
 4. Nope you don't have too many at all! State modeling is more a programming architecture question than a React thing. You'll get better at it overtime as you learn how complicated updating state can be when you've got a more complex data model.

File: src/Quiz.js [hunk 1]
```diff
@@ -1,208 +1,168 @@
...
-			JSON.stringify(incorrectCounter)
-		);
-	}
+			// should set state inside the async function according to : https://devtrium.com/posts/async-functions-useeffect
+			setQuestionsAll(allQuestions);
+			setQuestions(questionsFiltered);
+			setLoadingData(false);
+		};
+		getQuestions(); // todo need to catch this?
+	}, []); // when should useEffect be called??
```
Comment (alexrvincent) [inline src/Quiz.js:89]: I would rephrase the thinking to when should I *let* this run. useEffect is happy to run on every render if you let it. Your dependency array (the second parameter to useEffect aka your empty array) is the gatekeeper that prevents that from happening. 

So the answer to when should I *let* this run, is when you have something (usually a state variable) that was updated and you need to rerun this code with the new state variable. Let's say you needed to refetch the questions if a user swaps between practice mode and quiz mode. You'd pass one of the state variables there and determine what to do when it changes.

File: src/Quiz.js [hunk 2]
```diff
@@ -1,208 +1,168 @@
...
-			INCORRECT_COUNTER_LOCAL_STORAGE_KEY,
-			JSON.stringify(incorrectCounter)
-		);
-	}
+			// should set state inside the async function according to : https://devtrium.com/posts/async-functions-useeffect
+			setQuestionsAll(allQuestions);
+			setQuestions(questionsFiltered);
+			setLoadingData(false);
+		};
+		getQuestions(); // todo need to catch this?
```
Comment (alexrvincent) [inline src/Quiz.js]: I would for sure. Unhandled rejections are the worst errors to debug.

File: src/QuizProps.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import axios from 'axios';
```
Comment (alexrvincent) [inline src/QuizProps.js:1]: Using axios is definitely the right way to start things off and keep them simple. If you find yourself wanting a bit more data on the state of your API call (or some fancy things like refetching) I've been playing around with [React Query](https://react-query.tanstack.com/examples/simple) and it's been awesome (and really popular in the React community). Not suggesting any re-writes but something good to keep in mind for the future 😄

File: src/Quiz.js [hunk 4]
```diff
@@ -217,52 +177,45 @@ class Quiz extends Component {
 					{/* <i className="fa-solid fa-arrow-rotate-left" /> */}
 					<i className="fa-solid arrow-left" />
 					Finished all questions!
-					<button className="Quiz-restart-btn" onClick={this.resetQuestions}>
+					<button className="Quiz-restart-btn" onClick={() => resetGame(false)}>
```
Comment (alexrvincent) [inline src/Quiz.js]: Pro-tip: Defining anonymous arrow functions as callbacks actually creates a new function on every render 😮 . It's not a bug, but it's a slight performance issue if you only do it this way. I'd split this out into a `useCallback` and pass the callback in without an anonymous arrow function.

**Review by alexrvincent — CHANGES_REQUESTED** [review]

File: src/Quiz.js (hunk 2 above)
Comment (rszeredi) [reply src/Quiz.js]: Good call – logging the error now :)

File: src/Quiz.js (hunk 4 above)
Comment (rszeredi) [reply src/Quiz.js]: Thanks for the tip! I was a bit confused how to use `useCallback` here so ended up just defining separate named functions for each of the true/false cases. (Will keep `useCallback` in mind though)
